File: sdk/typescript/test/client.test.ts
import { describe, expect, it } from "bun:test";
import {
  RockskyAuthError,
  RockskyClient,
  RockskyHttpError,
  RockskyTimeoutError,
  createClient,
} from "../src";
import { lastUrl, mockFetch } from "./_mock";

describe("RockskyClient", () => {
  it("uses default base url when none provided", () => {
    const c = new RockskyClient();
    expect(c.config.baseUrl).toBe("https://api.rocksky.app");
  });

  it("strips trailing slash from base url", () => {
    const c = new RockskyClient({ baseUrl: "https://example.com/" });
    expect(c.config.baseUrl).toBe("https://example.com");
  });

  it("builds XRPC URL with namespace, method and query params", async () => {
    const { fetchImpl, calls } = mockFetch({ json: { handle: "alice" } });
    const c = createClient({
      baseUrl: "https://api.example",
      fetch: fetchImpl,
    });
    const r = await c.actor.getProfile({ did: "did:plc:abc" });
    expect(r).toEqual({ handle: "alice" });
    const url = lastUrl(calls);
    expect(url.pathname).toBe("/xrpc/app.rocksky.actor.getProfile");
    expect(url.searchParams.get("did")).toBe("did:plc:abc");
  });

  it("serializes array params with repeated keys", async () => {
    const { fetchImpl, calls } = mockFetch();
    const c = createClient({ fetch: fetchImpl });
    await c.graph.getFollowers({
      actor: "alice.bsky.social",
      dids: ["did:plc:1", "did:plc:2"],
    });
    const url = lastUrl(calls);
    expect(url.searchParams.getAll("dids")).toEqual([
      "did:plc:1",
      "did:plc:2",
    ]);
  });

  it("serializes boolean as 'true'/'false'", async () => {
    const { fetchImpl, calls } = mockFetch();
    const c = createClient({ fetch: fetchImpl });
    await c.scrobble.getScrobbles({ following: true });
    const url = lastUrl(calls);
    expect(url.searchParams.get("following")).toBe("true");
  });

  it("omits null/undefined params", async () => {
    const { fetchImpl, calls } = mockFetch();
    const c = createClient({ fetch: fetchImpl });
    await c.album.getAlbums({ limit: 10, genre: undefined });
    const url = lastUrl(calls);
    expect(url.searchParams.get("limit")).toBe("10");
    expect(url.searchParams.has("genre")).toBe(false);
  });

  it("sends procedures as POST with JSON body", async () => {
    const { fetchImpl, calls } = mockFetch({ json: { ok: true } });
    const c = createClient({ fetch: fetchImpl, auth: "TOKEN" });
    await c.scrobble.createScrobble({
      title: "Heart of Glass",
      artist: "Blondie",
    });
    const last = calls[calls.length - 1]!;
    expect(last.method).toBe("POST");
    expect(last.headers["content-type"]).toBe("application/json");
    expect(JSON.parse(last.body!)).toEqual({
      title: "Heart of Glass",
      artist: "Blondie",
    });
  });

  it("attaches bearer token from string auth", async () => {
    const { fetchImpl, calls } = mockFetch();
    const c = createClient({ fetch: fetchImpl, auth: "abc123" });
    await c.actor.getProfile();
    expect(calls[0]!.headers.authorization).toBe("Bearer abc123");
  });

  it("resolves async auth provider", async () => {
    const { fetchImpl, calls } = mockFetch();
    let count = 0;
    const c = createClient({
      fetch: fetchImpl,
      auth: async () => {
        count++;
        return `tok-${count}`;
      },
    });
    await c.actor.getProfile();
    await c.actor.getProfile();
    expect(calls[0]!.headers.authorization).toBe("Bearer tok-1");
    expect(calls[1]!.headers.authorization).toBe("Bearer tok-2");
  });

  it("throws RockskyAuthError if requireAuth without token", async () => {
    const { fetchImpl } = mockFetch();
    const c = createClient({ fetch: fetchImpl });
    await expect(c.scrobble.createScrobble({ title: "x", artist: "y" })).rejects.toBeInstanceOf(
      RockskyAuthError,
    );
  });

  it("withAuth returns a new client without mutating the original", async () => {
    const { fetchImpl, calls } = mockFetch();
    const c = createClient({ fetch: fetchImpl });
    const authed = c.withAuth("Z");
    await authed.actor.getProfile();
    await expect(c.scrobble.createScrobble({ title: "x", artist: "y" })).rejects.toBeInstanceOf(
      RockskyAuthError,
    );
    expect(calls[0]!.headers.authorization).toBe("Bearer Z");
  });

  it("throws RockskyHttpError on non-2xx with parsed body", async () => {
    const { fetchImpl } = mockFetch({ status: 404, json: { error: "nope" } });
    const c = createClient({ fetch: fetchImpl });
    try {
      await c.album.getAlbum({ uri: "at://x" });
      throw new Error("should have thrown");
    } catch (err) {
      expect(err).toBeInstanceOf(RockskyHttpError);
      const e = err as RockskyHttpError;
      expect(e.status).toBe(404);
      expect(e.body).toEqual({ error: "nope" });
    }
  });

  it("retries on 503 and succeeds", async () => {
    let n = 0;
    const { fetchImpl, calls } = mockFetch(() => {
      n++;
      if (n < 3) return { status: 503, text: "down" };
      return { json: { ok: true } };
    });
    const c = createClient({
      fetch: fetchImpl,
      retries: 3,
      retryDelayMs: 1,
    });
    const r = await c.actor.getProfile();
    expect(r).toEqual({ ok: true });
    expect(calls.length).toBe(3);
  });

  it("does not retry on 4xx", async () => {
    const { fetchImpl, calls } = mockFetch({ status: 400, json: {} });
    const c = createClient({
      fetch: fetchImpl,
      retries: 3,
      retryDelayMs: 1,
    });
    await expect(c.actor.getProfile()).rejects.toBeInstanceOf(RockskyHttpError);
    expect(calls.length).toBe(1);
  });

  it("times out and throws RockskyTimeoutError", async () => {
    const { fetchImpl } = mockFetch({ delayMs: 50, json: {} });
    const c = createClient({ fetch: fetchImpl, timeoutMs: 10 });
    await expect(c.actor.getProfile()).rejects.toBeInstanceOf(RockskyTimeoutError);
  });

  it("xrpc escape hatch hits arbitrary endpoint", async () => {
    const { fetchImpl, calls } = mockFetch({ json: { ok: true } });
    const c = createClient({ fetch: fetchImpl });
    await c.xrpc("app.rocksky.feed.search", "GET", {
      params: { query: "blondie" },
    });
    const url = lastUrl(calls);
    expect(url.pathname).toBe("/xrpc/app.rocksky.feed.search");
    expect(url.searchParams.get("query")).toBe("blondie");
  });
});

File: sdk/clojure/test/rocksky/client_test.clj
(ns rocksky.client-test
  (:require [cheshire.core :as json]
            [clojure.test :refer [deftest is testing]]
            [rocksky.client :as c]
            [rocksky.test-helpers :as h]))

(deftest client-defaults
  (let [client (c/client)]
    (is (= "https://api.rocksky.app" (:base-url client)))
    (is (= 30000 (:timeout-ms client)))
    (is (nil? (:token client)))))

(deftest client-trims-trailing-slash
  (is (= "http://localhost:3004"
         (:base-url (c/client {:base-url "http://localhost:3004/"})))
      "trailing slashes should be stripped from base-url")
  (is (= "http://localhost:3004"
         (:base-url (c/client {:base-url "http://localhost:3004///"})))
      "multiple trailing slashes are stripped"))

(deftest with-token-sets-bearer
  (let [client (-> (c/client) (c/with-token "tok-1"))]
    (is (= "tok-1" (:token client)))
    (is (= "tok-2" (:token (c/with-token client "tok-2")))
        "tokens replace, not stack")))

(deftest with-headers-merges
  (let [client (-> (c/client {:headers {"X-Trace" "1"}})
                   (c/with-headers {"X-User" "alice"}))]
    (is (= {"X-Trace" "1" "X-User" "alice"} (:headers client)))))

(deftest with-base-url-replaces
  (is (= "http://localhost:3004"
         (:base-url (-> (c/client) (c/with-base-url "http://localhost:3004/"))))))

(deftest query-sends-get-and-parses-json
  (let [[client calls] (h/mock-client {:response (h/json-response {:handle "alice"})})
        result (c/query client :app.rocksky.actor.getProfile {:did "did:plc:abc"})
        req    (h/last-call calls)]
    (is (= {:handle "alice"} result))
    (is (= :get (:method req)))
    (is (= "https://api.example.com/xrpc/app.rocksky.actor.getProfile" (:url req)))
    (is (= {"did" "did:plc:abc"} (:query-params req)))))

(deftest query-drops-nil-params
  (let [[client calls] (h/mock-client {})]
    (c/query client :app.rocksky.album.getAlbums {:limit 10 :offset nil :genre nil})
    (is (= {"limit" "10"} (:query-params (h/last-call calls)))
        "nil values must not show up in the query string")))

(deftest query-coerces-non-strings
  (let [[client calls] (h/mock-client {})]
    (c/query client :app.rocksky.scrobble.getScrobbles
             {:limit 50 :following true :did :alice})
    (is (= {"limit"     "50"
            "following" "true"
            "did"       "alice"}
           (:query-params (h/last-call calls))))))

(deftest query-supports-sequential-values
  (let [[client calls] (h/mock-client {})]
    (c/query client :app.rocksky.graph.getFollowers
             {:actor "alice" :dids ["did:1" "did:2"]})
    (let [params (:query-params (h/last-call calls))]
      (is (= "alice" (get params "actor")))
      (is (= ["did:1" "did:2"] (get params "dids"))))))

(deftest auth-header-added-when-token-present
  (let [[client calls] (h/mock-client {:token "secret"})]
    (c/query client :app.rocksky.actor.getProfile {:did "alice"})
    (is (= "Bearer secret"
           (get-in (h/last-call calls) [:headers "Authorization"])))))

(deftest auth-header-omitted-without-token
  (let [[client calls] (h/mock-client {})]
    (c/query client :app.rocksky.album.getAlbum {:uri "at://x"})
    (is (nil? (get-in (h/last-call calls) [:headers "Authorization"])))))

(deftest procedure-sends-post-with-json-body
  (let [[client calls] (h/mock-client {:response (h/json-response {:ok true})
                                       :token    "tok"})
        result (c/procedure client :app.rocksky.scrobble.createScrobble
                            {:title "Song" :artist "Artist"})
        req    (h/last-call calls)]
    (is (= {:ok true} result))
    (is (= :post (:method req)))
    (is (= "https://api.example.com/xrpc/app.rocksky.scrobble.createScrobble"
           (:url req)))
    (is (= :json (:content-type req)))
    (is (= {:title "Song" :artist "Artist"}
           (json/parse-string (:body req) true)))))

(deftest procedure-can-send-query-params-without-body
  (let [[client calls] (h/mock-client {})]
    (c/procedure client :app.rocksky.graph.followAccount nil {:account "bob"})
    (let [req (h/last-call calls)]
      (is (= :post (:method req)))
      (is (nil? (:body req)) "no JSON body when body arg is nil")
      (is (= {"account" "bob"} (:query-params req))))))

(deftest procedure-with-no-body-or-params
  (let [[client calls] (h/mock-client {})]
    (c/procedure client :app.rocksky.spotify.play)
    (let [req (h/last-call calls)]
      (is (= :post (:method req)))
      (is (nil? (:body req)))
      (is (nil? (:query-params req))))))

(deftest non-2xx-throws-with-ex-data
  (let [[client _] (h/mock-client {:response (h/json-response 404 {:error "NotFound"})})]
    (try
      (c/query client :app.rocksky.album.getAlbum {:uri "at://missing"})
      (is false "expected ex-info to be thrown")
      (catch clojure.lang.ExceptionInfo e
        (let [data (ex-data e)]
          (is (= 404 (:status data)))
          (is (= :app.rocksky.album.getAlbum (:nsid data)))
          (is (= :get (:method data)))
          (is (= {:error "NotFound"} (:body data))))))))

(deftest server-error-also-throws
  (let [[client _] (h/mock-client {:response (h/json-response 503 {:error "BadGateway"})})]
    (is (thrown? clojure.lang.ExceptionInfo
                 (c/procedure client :app.rocksky.scrobble.createScrobble
                              {:title "x" :artist "y"})))))
